### DISCUSSION TITLE
ObservedObjectの継承ってなんで必要なの？

### DISCUSSION THREAD

**Discussion by ikngtty** [sizumita/discord-ext-ui#7] [General]
あと ViewModel のプロパティに対する `Publised()` も。

メカニズムまだよく分かってないのでとりあえず聞いて学んでいくやつ。

**Comment by sizumita** [comment]
ObservedObjectを継承したオブジェクトをViewのインスタンス変数にすると、setattrしてるときにObservedobject.viewにselfを代入するという悪魔のような所業をしてる
で、Observedobjectのインスタンス変数が変更されたら、もしそれがPublishedでラップされてたら、view.updateを呼び出してアップデートされるという仕組み(

**Reply by ikngtty** [reply]
なるほど。ViewModel のプロパティ変更時に view.update を自動で呼びたいってことね。理解。

**Comment by sizumita** [comment]
ちなみにStateとかPublishedは、setattrの時に_watch_variables_みたいな変数に変数名が登録されて、中身が代入されるという仕様なので、これもかなりヤバい

**Reply by sizumita** [reply]
本当は型ヒントでできたら最高なんだけど、実行時には取れなさそうなので悲しい。

**Reply by ikngtty** [reply]
末尾までアンダーバーつけるの、見覚えない

**Reply by sizumita** [reply]
`__all__`とか、`__slots__`とかあるから見覚えしかない

**Reply by ikngtty** [reply]
ダブルアンダーバーはまた別じゃん

**Reply by ikngtty** [reply]
- `public_field`
- `_private_field`
- `__builtin_field__`

は見覚えあるけど、`_xxx_` は見覚えない。
＠Python詳しい人　あるの？

**Reply by sizumita** [reply]
ごめん、別に`_watch_variables`だった...(

**Reply by ikngtty** [reply]
うん。だよね。
